Add ComputerCraft integrations to more devices (1.20.1 edition)

ElementW opened the pull request:
> [!NOTE]  
> 1.20.1 rebase of #6849. I haven't probed extensively for breakage, so testing is more than welcome.

Adds a handful of ComputerCraft integrations to mostly train-related devices.
If merged I will expand the repo's wiki to document these APIs.

# <img src="https://github.com/user-attachments/assets/6fd6a74c-2509-447f-9df1-8a6a71e403c1" height="32" align="top"> Train Station

- Emit CC events `train_imminent`, `train_arrival` and `train_departure` when these occur on the station
  * 1 argument: train name
- Add CC `canTrainReach(destination)` function to test if the train stopped at the station can reach a (schedule-compatible) destination
- Add CC `distanceTo(destination)` function to get the shortest distance to a (schedule-compatible) destination at the moment, using the train stopped in the station

# <img src="https://github.com/user-attachments/assets/48440db4-0cef-4545-8ae8-cb53e80dffa4" height="32" align="top"> Train Signal

- Makes computer-controlled signals unable to be changed by external factors
- Emit CC event `train_signal_state_change` with the new state as parameter whenever the signal changes
- Add CC `getState()` function to get the current signal state
- Add CC `isForcedRed()`/`setForcedRed(forced)` functions to force the signal to be red, replacing redstone control
- Add CC `getSignalType()`/`cycleSignalType()` function to change the signal signal type in the same manner as using the wrench
- Add CC `listBlockingTrainNames()` function to list names of trains blocking the track group ahead of the signal

# <img src="https://github.com/user-attachments/assets/d3b1ee4c-e05c-4688-8a6d-07c01268526b" height="32" align="top"> Train Observer

- Emit CC events `train_passing` & `train_passed` with the name of the train originating these events
- Add CC `isTrainPassing()` function to check passing train presence
- Add CC `getPassingTrainName()` function to get the passing train name, or `nil` failing that

# <img src="https://github.com/user-attachments/assets/454bdc28-e43c-4e7f-8211-df6617202c6c" height="32" align="top"> Creative Motor

- Add CC `{get,set}GeneratedSpeed()` function get/set motor speed

# <img src="https://github.com/user-attachments/assets/aa207721-e517-466d-b4ba-cc43fb17bb36" height="32" align="top"> Sticker

- Add CC `isExtended()` function to check extension state
- Add CC `extend()`/`retract()`/`toggle()` to change the Sticker extension state, returning state change success

# <img src="https://github.com/user-attachments/assets/d6ef250a-4b5a-4cf1-b653-3c15f2bb1f25" height="32" align="top"> Speedometer

- Emit CC event `speed_change`
  - 1 argument: speed (RPM)

# <img src="https://github.com/user-attachments/assets/622bdece-5eb4-4fa2-9852-4a447ecf4cbb" height="32" align="top"> Stressometer

- Emit CC event `overstressed`
- Emit CC event `stress_change`
  - 2 arguments: stress (SU), capacity (SU)

# <img src="https://github.com/user-attachments/assets/b8c537e2-865e-43de-8c33-0b88a924858d" height="32" align="top"> Nixie Tubes

The big feature

- Makes computer-controlled Nixie Tubes unable to be changed by external factors (but can still be used as a Display Link source)
- Add CC `setText(text[, colour])` function
- Add CC `setTextColour(colour)` function
- Add CC `setSignal(first, second)` function taking 2 tables describing the appearance of the first and second tube as custom train signals
  * Has controls for Red, Green, Blue, glow dot size (1-4 pixels on each axis), and blink interval (in a PWM fashion)
    * `setSignal()` does partial display state update. Might be confusing, feedback welcome.

## Direct nixie control, no need for a display link

Now featuring colour!

https://github.com/user-attachments/assets/5d3b9128-19a7-477e-b082-14057491e06a

## Custom signals!

https://github.com/user-attachments/assets/2cc743d8-b8bc-45fb-9b9d-3ef021ab5f09

<details>
<summary>Code</summary>

Rotates the left tube through the Oklch hue wheel, and the right tube through the valid glow sizes

```lua
function oklch2oklab(l, c, h)
    if h ~= h then -- Checks for NaN
        return l, 0, 0
    end
    return  l, c * math.cos(h * math.pi / 180), c * math.sin(h * math.pi / 180)
end
function oklab2rgb(l, a, b)
    -- Convert OKLab to linear RGB
    local l_ = l + 0.3963377774 * a + 0.2158037573 * b
    local m_ = l - 0.1055613458 * a - 0.0638541728 * b
    local s_ = l - 0.0894841775 * a - 1.2914855480 * b

    local l_3 = l_ ^ 3
    local m_3 = m_ ^ 3
    local s_3 = s_ ^ 3

    local r = 4.0767416621 * l_3 - 3.3077115913 * m_3 + 0.2309699292 * s_3
    local g = -1.2684380046 * l_3 + 2.6097574011 * m_3 - 0.3413193965 * s_3
    local b = -0.0041960863 * l_3 - 0.7034186147 * m_3 + 1.7076147010 * s_3

    -- Convert linear RGB to sRGB
    local function linearToSrgb(c)
        if c <= 0.0031308 then
            return 12.92 * c
        else
            return 1.055 * (c ^ (1 / 2.4)) - 0.055
        end
    end

    r = linearToSrgb(r)
    g = linearToSrgb(g)
    b = linearToSrgb(b)

    -- Clamp values to [0, 1] range
    r = math.min(math.max(r, 0), 1)
    g = math.min(math.max(g, 0), 1)
    b = math.min(math.max(b, 0), 1)

    return r * 255, g * 255, b * 255
end

t=peripheral.wrap("top")
i=0
gw = {
    {1, 1},
    {1, 2},
    {1, 3},
    {1, 4},
    {2, 4},
    {3, 4},
    {4, 4},
    {4, 3},
    {4, 2},
    {4, 1},
    {3, 1},
    {2, 1},
}
while true do
    i = (i + 4) % 360
    local r, g, b = oklab2rgb(oklch2oklab(0.5, 0.23, i))
    local j = math.floor(i/16) % #gw
    t.setSignal({r=r, g=g, b=b, glowWidth=1, glowHeight=1}, {glowWidth=gw[j+1][1], glowHeight=gw[j+1][2]})
    sleep(0.05)
end
```

</details>

## Clears & reset themselves when dis/connected from computers

https://github.com/user-attachments/assets/3f84b17e-76bb-4010-a5c9-20d527f14d9a

ElementW committed:
Add ComputerCraft integration to Nixie Tubes

- Makes computer-controlled Nixie Tubes unable to be changed by external
  factors (but can still be used as a Display Link source)
- Add CC setText(text[, colour]) function
- Add CC setTextColour(colour) function
- Add CC setSignal(first, second) function taking 2 tables describing
  the appearance of the first and second tube as custom train signals

ElementW committed:
Add ComputerCraft integration to Train Signals

- Makes computer-controlled signals unable to be changed by external
  factors
- Emit CC event train_signal_state_change with the new state as
  parameter whenever the signal changes
- Add CC getState() function to get the current signal state
- Add CC isForcedRed()/setForcedRed(forced) functions to force the
  signal to be red, replacing redstone control
- Add CC getSignalType()/cycleSignalType() function to change the
  signal signal type in the same manner as using the wrench
- Add CC listBlockingTrainNames() function to list names of trains
  blocking the track group ahead of the signal

ElementW committed:
Improve ComputerCraft integration of Train Stations

- Emit CC events train_imminent, train_arrival and train_departure when
  these occur on the station
- Add CC canTrainReach() function to test if the train stopped at the
  station can reach a (schedule-compatible) destination
- Add CC distanceTo() function to get the shortest distance to a
  (schedule-compatible) destination at the moment, using the train
  stopped in the station

ElementW committed:
Add ComputerCraft integration to Train Observers

- Emit CC events train_passing & train_passed with the name of the train
  originating these events
- Add CC isTrainPassing() function to check passing train presence
- Add CC getPassingTrainName() function to get the passing train name,
  or nil failing that

ElementW committed:
Add ComputerCraft integration to Creative Motors

- Add CC {get,set}GeneratedSpeed() function get/set motor speed

ElementW committed:
Add ComputerCraft integration to Stickers

- Add CC isExtended() function to check extension state
- Add CC extend()/retract()/toggle() to change the Sticker extension
  state, returning state change success.

ElementW committed:
Emit ComputerCraft events from Speedometers/Stressometers

- Emit CC event `overstressed` from Stressometers
- Emit CC event `stress_change` from Stressometers
  - 2 arguments: stress (SU), capacity (SU)
- Emit CC event `speed_change` from Speedometers
  - 1 argument: speed (RPM)

Ordep-42 commented:
Very nice integration that opens a lot of possibilites, really hope it gets merged into the main mod!

BirbIrl commented:
Very cool and would love to see it merged! i'm also currently working on a PR with CC integration for the StockTicker and packagers. Are you planning to work on that as well? You seem way better than me at this so maybe i should just leave the stage entirely for you. I barely managed to add tickers as new peripherals that just print any value after a good hour of bashing my head against CC's source

ElementW commented:
@BirbIrl I'm not planning on adding support for the new logistics stuff (as least yet, ADHD is a bitch and who knows I might still end up adding new things). Feel free to base your work on mine however, the `ComputerEvent` related code is probably what's going to be the most useful to you.

BirbIrl commented:
> @BirbIrl I'm not planning on adding support for the new logistics stuff (as least yet, ADHD is a bitch and who knows I might still end up adding new things). Feel free to base your work on mine however, the `ComputerEvent` related code is probably what's going to be the most useful to you.

Aye, going back to work on it (when my brain lets me, aspergers is a bitch), I'll try making it with your pr getting merged in mind 🫡

samuelh2005 commented:
I would like a function to get a station's timetable similar to how the Display Link / Board can do:

![2025-03-12_14 21 51](https://github.com/user-attachments/assets/73ad088d-43c7-4182-91b5-f1b458cea14b)

This will allow me to create a much smaller version by using CC monitors.

---

This could be like:

`trainStation.getTimetable()` >

```lua
{
    { arrivesIn = 0, name = "Cargo 1", destination = "Factory Land", departsIn = 10 }, -- 0 means the train is at the station.
    { arrivesIn = 1200, name = "Banana Express", destination = "Banana Land", departsIn = nil }, -- nil is used becuase the train has not arrived yet.
    { arrivesIn = 200, name = "Oak Train", destination = "The Forest", departsIn = nil }
}
```

tizu69 commented:
@ajh123 great to see someone else asking for this :p

What would this return for `~`, so when it is not aware of its time?

ElementW commented:
> I would like a function to get a station's timetable similar to how the Display Link / Board can do

I've wanted to do this when I made the code a few months ago but then realised it was going to be more complex than anticipated. Maybe I'll revisit this, but again, no promises.

In fact I wanted to give CC access to the full train, track & junction graph of a train network, but it's a whole endeavour onto itself and I only have a limited motivation span, to my chagrin.

svemat01 commented:
Awesome work! Really looking forward to seeing this get merged!

samuelh2005 commented:
I've made a pull request https://github.com/Creators-of-Create/Create/pull/8491 which allows computers to get more information about a train.

Whilst it isn't the full display board information or the whole track graph it seems like a good starting point.

CrypticCubed commented:
First time comment on Github so forgive any noobiness or rudeness, but is this merge or any further ComputerCraft compatibility with train-related blocks still not being considered for any reason? Currently trying to achieve a live-feed for train positions but ComputerCraft integration is still really limited; only stations are a peripheral so we can't pinpoint trains along large distances of track unless we make some assumptions with redstone data from train signals/observers..

Addons offering this much integration are sparse and incompatible but now that I've found it's already being proposed & provided for Create itself, this just seems like a trivial choice for at _least_ some better integration with ComputerCraft and Create's train system, alongside everything else included in this pull request. 🙂

samuelh2005 commented:
> Currently trying to achieve a live-feed for train positions but ComputerCraft integration is still really limited; only stations are a peripheral so we can't pinpoint trains along large distances of track unless we make some assumptions with redstone data from train signals/observers..

There is the [CC:C Bridge](https://modrinth.com/mod/cccbridge) mod which can read display link data (which means you can retreive a complete timetable for a station) but unfornatuantly it does not pinpoint the excat location of a train. 

For excat positioning you could install the [Create Track Map](https://modrinth.com/mod/create-track-map) and interface with the web map with CC by using the HTTP API.

BirbIrl commented:
> but is this merge or any further ComputerCraft compatibility with train-related blocks still not being considered for any reason? 

The CC x Create Update 6 pr took like 4 months of active development/review to officially merge (still on the dev branch), which is still somehow faster than this pr since this was originally made exactly a year ago. I'm pretty sure for the longest while now the maintainers have been busy with fixing the things they've introduced in update 6 and only now slowly moving on to looking at pre update 6 features to tweak/update, so i don't think this'll get closed, just gonna take a while longer to merge.

I'm looking forward to that as well, this pr also brings a really nice set of utilities for adding events to future blocks,

BirbIrl committed:
Merge remote-tracking branch 'ElementW/mc1.20.1/dev' into mc1.20.1/ElementWIntegrations

BirbIrl commented:
Alright i'm available again

got this working for the current dev build [here](https://github.com/BirbIrl/Create/tree/mc1.20.1/ElementWIntegrations). It only required rewriting like two lines, but i found a bug regarding nixie tubes, they don't reset if you give them a custom ComputerSignal and detach a pc. Gonna add a few opinionated commits before submitting a PR on your repo @ElementW, and then you can pull it in however you wish, i'll try and make the opinionated commits small and easy to read :D

I *might* also add events to the new peripherals from our logistics integration pr.

I *might* also add these peripherals to the [serial tests world](https://github.com/BirbIrl/Create-CC-Integration-Tests), but i'll probably burn out by then.

I *will* update the [wiki PR](https://github.com/Creators-of-Create/wiki/pull/4) to include this PR as well (since i'm like 99% sure this will get pulled in)

BirbIrl committed:
drops the computerSignal on disconnect.

Leaving it with the signal is kinda useful for making pretty builds, but
it breaks when you put nixie tubes in series, it just kinda overrides it
in a very unnatural way and doesn't really coorelate with how text is rendered.

BirbIrl commented:
Yo @ElementW your code for checking if the signal block is still connected to the computer is very... weird(?)
I'm not quite sure i understand it, but, it only seems to drop the setForcedRed override when you break the pc next to it. This means if you power the pc off/break the modem next to it, it's stuck in the setForcedRed state.

Hitting you up in case you instantly know how to fix it. Here's the relevant lines of code: 
- [SignalBlock.java](https://github.com/BirbIrl/Create/blob/448222e09f87360acb1558a0ff3d67f252c1b848/src/main/java/com/simibubi/create/content/trains/signal/SignalBlock.java#L99)
- [SignalBlockEntity.java](https://github.com/BirbIrl/Create/blob/448222e09f87360acb1558a0ff3d67f252c1b848/src/main/java/com/simibubi/create/content/trains/signal/SignalBlockEntity.java#L165)

There seeems to be some block vs blockentity communicating weirdness that i'm not getting here. I'm too skill issued on the minecraft engine knowledge to understand this :P

BirbIrl committed:
added `isAttachedToBlock` to sticker peripheral

also needs to check if it's extended because it'll prock yes if there's
a block in front, but it's not extended to grab it. Should be more
intuitive.

BirbIrl commented:
My job here is done :>

Merges cleanly, patched a smol thing and added another thing in the PR on ElementW's end (https://github.com/ElementW/Create/pull/1)

Made a wiki PR, documenting all the new peripherals and events (https://github.com/Creators-of-Create/Create/pull/7453)

Didn't end up adding the events for the packager/repackager/redstone requester/stock ticker/frogpost/postbox/ or making an update to the test world because the big lazy hath struck me.

It Works :tm:™ and is Documented :tm:™

ElementW committed:
Merge pull request #1 from BirbIrl/mc1.20.1/ElementWIntegrations

merged with upstream + nixie tweak

BirbIrl commented:
Threw on a PR to @ElementW 's fork, here: https://github.com/ElementW/Create/pull/2. If it gets merged on their end it should pop up here

added events to the logistics blocks. Currently there's no easy way to check what package a packager just made, and what's inside the package, so the new `package_created` event should be plenty useful, letting you peek inside the package object.

Added other events as well while i was at it.

BirbIrl committed:
Added package events to logistics peripherals

events added to the packager, repackager, frogport and postbox

rebased and cleaned

IThundxr commented:
Thanks!
